### DISCUSSION TITLE
livewire/message{name} return 404 when in subfolder install

### DISCUSSION THREAD

**Discussion by marcdegagne** [livewire/livewire#5731] [1. Help]
Hey all,

I have a project I just migrated to production and now all components methods that responds to an event like click or update, I get a 404 popup and in the browser console I get: `https://domaine_name.com/subdir/livewire/message/show-image-module 404 ` just to name this one.
All mount methods are working as expected.

What I am experiencing is very similar to https://github.com/livewire/livewire/discussions/5557 which does not have any conclusion.

Running on shared hosting apache/cPanel with Laravel 8.74 and livewire 2.8.

**Here is the structure of the live project:**

 - /home/user/laravel_app
 -- all Laravel  folders and files
 
- /home/user/public_html/
-- Virtual subdir using a symbolic link to /home/user/laravel_app/public

Probably related to the new struture after migration.
All images and scripts are loading after properly configurating the different paths and configs.

**.env**
`APP_URL=https://my-domain.com`
`ASSET_URL=/subdir`

**config**
livewire.php
'asset_url' => config( "app.asset_url")`,

Did all artisan commands to clear config, route, cache and views.

Searched and testeds some hints but so far ziltch! I am running out of ideas.

Thanks!

**Comment by marcdegagne** [comment]
How to make it work in a subfolder ?

 `POST | livewire/message/{name} |  livewire.message  | Livewire\Controllers\HttpConnectionHandler`

Any ideas?

**Comment by tomphilpotts** [comment]
I am also having the same issue. 

Managed to get everything else working through a reverse proxy. 

But all post requests are returning 404.

**Reply by tomphilpotts** [reply]
` 'asset_url' => env('ASSET_URL', null),`
`'app_url' => env('APP_URL', null),`

**Reply by marcdegagne** [reply]
Hey Tom @tomphilpotts,

Do you have any advancement on this issue?
Nothing so far on my end at this point. I am about to try a different installation structure to see if I can get over this problem.

**Comment by marcdegagne** [comment]
Have managed to install using a different structure as before and still the same thing, 404.

**I have on my /home/user/:**
**laravel_app**
-- pulbic and all the rest

**public_html**
--symbolic link to point to laravel_app/public

Everthing is working eccept for the Livewire POST requests. No change.

**Reply by marcdegagne** [reply]
Can't figure it out!
All other livewire route listed in route:list seams to work.

I tried to `dd("something");` within the function that is being called on click and it does'nt event get to the function. Nothing prints.
I changed the function to not use any parameters thinking it could be the model binding giving an error. Still 404.

**Reply by tomphilpotts** [reply]
Hi

I managed to get it working. I am using a reverse_proxy pointing to an different IP. 

This is the code I used to get it working.

```
  location /folder/ {
      proxy_pass http://IP_ADDRESS:80/;
      proxy_http_version 1.1;
      proxy_buffering off;
      proxy_cache off;
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto https;
      proxy_set_header X-Forwarded-Path /directory;
      proxy_set_header X-Forwarded-Prefix "/directory";
      proxy_pass_header X_LIVEWIRE;
      proxy_pass_header X-CSRF-TOKEN;
  }
```

- I learnt that I need the slash at the end of the IP and the folder.
- Also publish the vendor files for Livewire. Add them to composer: https://laravel-livewire.com/docs/2.x/installation#publish-assets

**Reply by marcdegagne** [reply]
Thanks Tom!

This reverse proxy is out of my league. I will keep that solution as last resort.

**Comment by marcdegagne** [comment]
Could this be related to the fact that Livewire (before v 3.* at least) does not support custom configurable routes?
See https://github.com/livewire/livewire/pull/2662 PR.

**Comment by daniftinca** [comment]
Hi, not sure if you managed to fix it but I encountered the same issue now for an app that has been running in production for 2 years now after upgrading to laravel 10 (and thus automatically upgrading to the latest livewire v2 version). 

I haven't fixed it yet, but I identified the problem and it is stemming from this File:  vendor/livewire/livewire/src/Controllers/HttpConnectionHandler.php

My issue was with the following lines: 
![image](https://github.com/livewire/livewire/assets/17156755/078cf988-ab17-4d71-b04e-26b02d302589)

Once I removed them it all worked well.

**Reply by marcdegagne** [reply]
Hey Dan,

No fix beside publishing the site as subdomains. One for the admin and another for the frontend.
So now all of requests are at the root of each.
```php
Route::domain('admin.thedomain.com' )->group( function () {
  // auth
}
Route::domain('fontend.thedomain.com' )->group( function () {
  // frontend path
}

```

**Comment by wizzymore** [comment]
You are showing this: `APP_URL=https://my-domain.com` the problem might be from here, you need to update the url to include the sub_folder

**Comment by noktius** [comment]
Hi everyone, I faced the same problem and I've spent 3 days trying to find any possible solution for this problem, but nothing works and I really would like to have any possible idea how to fix it..

So I have **a main website (example.com)** and **a Laravel/Livewire forum (example.com:8081)**. Both in **different** Dockers, but on **the same network**. The site is on 80 port, the forum is on 8081. 

But I need to have the forum being accessed with **example.com/forum**.

Using **nginx** and **proxy_pass**, I had the following thing:
```
location /forum/ {
        proxy_pass http://IP:8081;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        add_header 'Access-Control-Allow-Origin' '*';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,x-csrf-token,x-livewire';
        add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range';
        rewrite ^/forum/(.*)$ /$1 break;
    }
```

Everything works find besides.. **built-in Livewire POST-request in JavaScript** (ex. Notifications). I always get 
`index.js:46     POST http://example.com/forum/livewire/message/discussions 404 (Not Found)`

Using Postman, it works perfectly and I always receive the message "_CSRF token mismatch_.", so fine. So... why????.....

I tried basically everything:
1. Every possible combination of `api_url` and `asset_url`
2. Every possible combination of` 'path' => 'forum/livewire'` (does it exist in livewire.php?)
3. Some crazy things like` location ~ ^/livewire/(.*) { proxy_pass http://IP:8081/livewire/$1;`
4. I tried even `location /forum/livewire/message/discussions { proxy_pass http://IP:8081/livewire/message/discussions;` !!!! (really crazy)
...


But I'm really stuck. This problem has stopped my development for 3 days already :-(

I tried the solution from @tomphilpotts but I don't understand which directory should I put into `proxy_set_header X-Forwarded-Path /directory;`
(my **docker-compose**:  `volumes: - ./src:/var/www/app`)
I tried `/var/www/app`, `/app`, `/forum` - nothing worked.....

So it was like:
```
location /forum/ {
        proxy_pass http://IP:8081/;
        proxy_http_version 1.1;
      proxy_buffering off;
      proxy_cache off;
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto http;
      proxy_set_header X-Forwarded-Path /app;
      proxy_set_header X-Forwarded-Prefix "/app";
      proxy_pass_header X_LIVEWIRE;
      proxy_pass_header X-CSRF-TOKEN;
    }
```
I would try a solution from @daniftinca but I really don't want to mess up with Livewire src...

Anyway I'm super confused, why Postman works and JS-POST - no. That's really strange for me!

Thanks in advance for any kind of help :-)....
